Implement dynamic scaling of the LineEdit right icon based on control size and scale factor

nongvantinh opened the pull request:
Close: https://github.com/godotengine/godot-proposals/issues/10493

The implementation allows the LineEdit node to scale the right icon to match the font size first. Then, when the `expand_icon` option is enabled, the icon will expand to the full height of the node. The scale of the icon can then be controlled using the scale factor.


https://github.com/user-attachments/assets/bed11cde-9424-4f0f-9d6a-d7145eb40dda


<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

KoBeWi commented on doc/classes/LineEdit.xml:
```diff
@@ -275,6 +275,9 @@
 			If [code]false[/code], existing text cannot be modified and new text cannot be added.
 		</member>
+		<member name="expand_icon" type="bool" setter="set_expand_icon" getter="is_expand_icon" default="false">
+			When enabled, the button's icon will expand/shrink to fit the LineEdit's size.
```
You could mention what happens when this is disabled. The icon will be resized to text size.

KoBeWi commented on doc/classes/LineEdit.xml:
```diff
@@ -321,6 +324,9 @@
 			Sets the icon that will appear in the right end of the [LineEdit] if there's no [member text], or always, if [member clear_button_enabled] is set to [code]false[/code].
 		</member>
+		<member name="right_icon_scale" type="float" setter="set_right_icon_scale" getter="get_right_icon_scale" default="1.0">
+			Used with the expand_icon to control the icon size.
```
```suggestion
			Scale ratio of the icon when [member expand_icon] is enabled.
```

KoBeWi commented:
This changes the default behavior of the right icon. It will scale to text, while previously it would use original size.
I think the behavior should be preserved. You can change `expand_icon` to `expand_mode` enum and allow all 3 behaviors (original size, fit to text and fit to LineEdit).

nongvantinh committed:
Implement dynamic scaling of the LineEdit right icon based on control size and scale factor

    The implementation allows the LineEdit node to scale the right icon to match the font size first. Then, when the `expand_icon` option is enabled, the icon will expand to the full height of the node. The scale of the icon can then be controlled using the scale factor.

Co-authored-by: Tomasz Chabora <kobewi4e@gmail.com>
Co-authored-by: A Thousand Ships <96648715+AThousandShips@users.noreply.github.com>

KoBeWi commented on doc/classes/LineEdit.xml:
```diff
@@ -276,6 +276,9 @@
 			If [code]false[/code], existing text cannot be modified and new text cannot be added.
 		</member>
+		<member name="expand_mode" type="int" setter="set_expand_mode" getter="get_expand_mode" enum="LineEdit.ExpandMode" default="0">
+			Define the scaling behavior of the right icon
```
```suggestion
			Define the scaling behavior of the [member right_icon].
```

KoBeWi replied:
```suggestion
			Define the scaling behavior of the [member right_icon].
```

KoBeWi commented on doc/classes/LineEdit.xml:
```diff
@@ -322,6 +325,9 @@
 			Sets the icon that will appear in the right end of the [LineEdit] if there's no [member text], or always, if [member clear_button_enabled] is set to [code]false[/code].
 		</member>
+		<member name="right_icon_scale" type="float" setter="set_right_icon_scale" getter="get_right_icon_scale" default="1.0">
+			Scale ratio of the icon when [member expand_mode] is set to fit to LineEdit.
```
```suggestion
			Scale ratio of the icon when [member expand_mode] is set to [constant EXPAND_MODE_FIT_TO_LINE_EDIT].
```

KoBeWi commented on scene/gui/line_edit.cpp:
```diff
@@ -352,6 +352,41 @@ void LineEdit::_delete(bool p_word, bool p_all_to_right) {
+Point2 LineEdit::_get_right_icon_size(Ref<Texture2D> p_right_icon, bool p_get_min) const {
+	Size2 icon_size;
+
+	if (!p_right_icon.is_valid()) {
```
```suggestion
	if (p_right_icon.is_null()) {
```

KoBeWi commented on scene/gui/line_edit.cpp:
```diff
@@ -352,6 +352,41 @@ void LineEdit::_delete(bool p_word, bool p_all_to_right) {
+		return icon_size;
+	}
+
+	ExpandMode mode = p_get_min ? EXPAND_MODE_FIT_TO_TEXT : expand_mode;
```
Why this condition? It's different from original behavior.

https://github.com/user-attachments/assets/1cf83ee7-59dc-4608-a47a-771464c73683

Old:

https://github.com/user-attachments/assets/38e4d538-65bf-4f23-80d0-cfa9fa2cd83c

The icon never leaves the LineEdit.

nongvantinh replied:
I want the LineEdit to be scaled down as small as possible. That's why it's there. If this changes the default behavior, it should be removed.

KoBeWi commented on scene/gui/line_edit.cpp:
```diff
@@ -2052,12 +2093,14 @@ Size2 LineEdit::get_minimum_size() const {
-		icon_max_width = MAX(icon_max_width, theme_cache.clear_icon->get_width());
+		Point2 right_icon_size = _get_right_icon_size(theme_cache.clear_icon);
+		min_size.height = MAX(min_size.height, right_icon_size.height);
+		icon_max_width = MAX(icon_max_width, right_icon_size.width);
```
If right icon is bigger, the minimum size will fit to the right icon, even if it's not used when clear button is enabled.
It's a pre-existing bug, so you don't need to fix it now.

KoBeWi commented on scene/gui/line_edit.cpp:
```diff
@@ -2523,6 +2566,31 @@ Ref<Texture2D> LineEdit::get_right_icon() {
 }
 
+void LineEdit::set_icon_expand_mode(ExpandMode p_mode) {
+	if (icon_expand_mode != p_mode) {
```
```suggestion
	if (icon_expand_mode = p_mode) {
		return;
	}
```
Early returns are preferred. Same below.

KoBeWi reviewed (approved):
Overall looks good now.

AThousandShips commented on scene/gui/line_edit.cpp:
```diff
@@ -2546,6 +2589,35 @@ Ref<Texture2D> LineEdit::get_right_icon() {
+}
+
+void LineEdit::set_right_icon_scale(float p_scale) {
+	if (right_icon_scale != p_scale) {
```
```suggestion
	if (right_icon_scale == p_scale) {
```

nongvantinh commented:
This PR is older than my entire volatile memory.

Repiteo commented:
Thanks!
